File: ProjectTracker10/ProjectTracker.Library/ResourceList.vb
Imports System.Data.SqlClient
Imports CSLA.Data

<Serializable()> _
Public Class ResourceList
  Inherits CSLA.ReadOnlyCollectionBase

#Region " Data Structure "

  <Serializable()> _
  Public Structure ResourceInfo
    ' this has private members, public properties because
    ' ASP.NET can't databind to public members of a structure...
    Private mID As String
    Private mName As String

    Public Property ID() As String
      Get
        Return mID
      End Get
      Set(ByVal Value As String)
        mID = Value
      End Set
    End Property

    Public Property Name() As String
      Get
        Return mName
      End Get
      Set(ByVal Value As String)
        mName = Value
      End Set
    End Property

    Public Overloads Function Equals(ByVal info As ResourceInfo) As Boolean
      Return mID = info.ID
    End Function
  End Structure

#End Region

#Region " Business Properties and Methods "

  Default Public ReadOnly Property Item(ByVal Index As Integer) As ResourceInfo
    Get
      Return CType(list.Item(Index), ResourceInfo)
    End Get
  End Property

#End Region

#Region " Contains "

  Public Overloads Function Contains( _
    ByVal item As ResourceInfo) As Boolean

    Dim child As ResourceInfo

    For Each child In list
      If child.Equals(item) Then
        Return True
      End If
    Next

    Return False

  End Function

#End Region

#Region " Shared Methods "

  Public Shared Function EmptyList() As ResourceList
    Return New ResourceList
  End Function

  Public Shared Function GetResourceList() As ResourceList
    Return CType(DataPortal.Fetch(New Criteria), ResourceList)
  End Function

  <Serializable()> _
  Public Class Criteria
    ' no criteria - we retrieve all resources
  End Class

  Protected Sub New()
    ' prevent direct creation
  End Sub

#End Region

#Region " Data Access "

  Protected Overrides Sub DataPortal_Fetch(ByVal Criteria As Object)
    Dim cn As New SqlConnection(db("PTracker"))
    Dim cm As New SqlCommand

    cn.Open()
    Try
      With cm
        .Connection = cn
        .CommandType = CommandType.StoredProcedure
        .CommandText = "getResources"

        Dim dr As New SafeDataReader(.ExecuteReader)
        Try
          While dr.Read()
            Dim info As ResourceInfo
            info.ID = dr.GetString(0)
            info.Name = dr.GetString(1) & ", " & dr.GetString(2)
            innerlist.Add(info)
          End While

        Finally
          dr.Close()
        End Try

      End With

    Finally
      cn.Close()
    End Try
  End Sub

#End Region

End Class

File: ProjectTrackercs10/ProjectTracker.Library/ResourceList.cs
using System;
using System.Data;
using System.Data.SqlClient;
using CSLA;
using CSLA.Data;

namespace ProjectTracker.Library
{
  [Serializable()]
  public class ResourceList : ReadOnlyCollectionBase
  {
    #region Data Structure

    [Serializable()]
      public struct ResourceInfo
    {
      // this has private members, public properties because
      // ASP.NET can't databind to public members of a structure...
      private string _id;
      private string _name;

      public string ID
      {
        get
        {
          return _id;
        }
        set
        {
          _id = value;
        }
      }

      public string Name
      {
        get
        {
          return _name;
        }
        set
        {
          _name = value;
        }
      }

      public bool Equals(ResourceInfo info)
      {
        return _id == info.ID;
      }
    }

    #endregion

    #region Business Properties and Methods

    public ResourceInfo this [int index]
    {
      get
      {
        return (ResourceInfo)List[index];
      }
    }

    #endregion

    #region Contains

    public bool Contains(ResourceInfo item)
    {
      foreach(ResourceInfo child in List)
        if(child.Equals(item))
          return true;
      return false;
    }

    #endregion

    #region Static Methods

    public static ResourceList EmptyList()
    {
      return new ResourceList();
    }

    public static ResourceList GetResourceList() 
    {
      return (ResourceList)DataPortal.Fetch(new Criteria());
    }

    #endregion

    #region Criteria 

    [Serializable()]
      private class Criteria
    {
      // no criteria - we retrieve all resources
    }

    #endregion

    #region Constructors

    private ResourceList()
    {
      // prevent direct creation
    }

    #endregion

    #region Data Access

    protected override void DataPortal_Fetch(object criteria)
    {
      using(SqlConnection cn = new SqlConnection(DB("PTracker")))
      {
        cn.Open();
        using(SqlCommand cm = cn.CreateCommand())
        {
          cm.CommandType = CommandType.StoredProcedure;
          cm.CommandText = "getResources";

          using(SafeDataReader dr = new SafeDataReader(cm.ExecuteReader()))
          {
            locked = false;
            while(dr.Read())
            {
              ResourceInfo info = new ResourceInfo();
              info.ID = dr.GetString(0);
              info.Name = dr.GetString(1) + ", " + dr.GetString(2);
              List.Add(info);
            }
            locked = true;
          }
        }
      }
    }

    #endregion

  }
}
